### ISSUE TITLE
StringIndexOutofBoundsException on malformed url in org.eclipse.jetty.http.HttpURI

### ISSUE THREAD

**Issue by dkomanov** [eclipse/jetty.project#876]
This code:

```
new HttpURI("http://www.example.com/#?").getQuery()
```

Will throw an exception:

```
java.lang.StringIndexOutOfBoundsException: String index out of range: -2
    at java.lang.String.checkBounds(String.java:381)
    at java.lang.String.<init>(String.java:462)
    at org.eclipse.jetty.http.HttpURI.getQuery(HttpURI.java:709)
```

The actual problem is when you get such invalid request in Jetty, the `HttpServletRequest#getQueryString` throws an exception, which is unexpected.

**Comment by joakime** [comment]
Interesting.

The raw `#` makes everything after it part of the URI fragment identifier.
Attempting to access .getQuery() on that should fail, but not with a StringIndexOutOfBoundsException.

Why do you need the `#` in your URI in a raw (unencoded) form?  

That's not a character that's valid in a URI that Jetty would ever see/use normally. (A server cannot use it to process a request, and a User-Agent is forbidden from sending it.)

**Comment by dkomanov** [comment]
I'm really sorry, I should check this in advance. Affected version is `9.1.6.v20160112` (and at least `9.1.2.v20140210`, so, I assume, all 9.1.x versions). In 9.3 everything looks fine.

I think, it was fixed in this commit: https://github.com/eclipse/jetty.project/commit/30123607c66627f74c8e8f7bed372f6e9b99ec31.

**Comment by joakime** [comment]
Considering this fixed in 9.3+
